File: Concordia_COMP_6411_COMPARAT_STUDY_PROGRAM_LANG/diffCode/Python/potum.py
# -*- coding: utf-8 -*-
##############################################################################

__author__      = "Safan Maredia"
__copyright__   = "Copyright 2013, POTUM: THE BEVERAGE COASTER PROJECT"
__license__     = "GPL"
__version__     = "1.1.0"
__email__       = "marediasafan@gmail.com"
__status__      = "Production"
__description__ = """Module calculating value of l for which area of overlapping coaster
                  is half the area of any two of the coaster"""

##############################################################################

from decimal import Decimal
import math
import constants

#get power number upto which series needs to be solved based on provided accuracy
def calculatePowerNum(accuracy):
    pi = constants.calculatePi(100)
    powerNum = 3                            #series starts with power of 3
    for powerNum in range(3,1000,2):        #range from 3 to 1000 with a step size=2
        if pi**powerNum/(constants.factorial(powerNum)) < accuracy:
            break 

    return powerNum                

#expand the series upto powerNum to get the value of alpha based on provided accuracy
def calculateAlpha(accuracy):
    powerNum = calculatePowerNum(accuracy)
    pi = constants.calculatePi(100)

    times = 1                             #counter to track iteration performed for getting value of alpha 
    alpha = 2*pi/3                        #lower bound of alpha

    while alpha < pi:
        sum = 0
        for x in range(1, int((powerNum-1)/2)):
            sum += (-1)**(x+1)*(alpha**(2*x+1)) / constants.factorial(2*x+1)

        if(abs(Decimal(pi/2) - Decimal(sum))) < accuracy:
            break
                        
        alpha = alpha + accuracy
        times = times + 1

    return alpha

#get value of l for a certain value of r
def calculateLength(accuracy, r):
    R=r
    pi    = constants.calculatePi(100)
    alpha = calculateAlpha(accuracy)

    #calculating trignometric functions    
    sinAlpha = alpha-pi/2
    cosAlpha = 0
    cosAlpha = -math.sqrt(1-sinAlpha*sinAlpha)
    cosHalfAlpha = math.sqrt((1+cosAlpha)/2)

    valueL = 2 * R * Decimal((1-cosHalfAlpha))
                
    #Print the output
    print ("\nCalculated Pi value: ", pi)
    print ("\nCalculated Alpha value: ", alpha)
    print ("\nLength of Overlap of the two circles (L): ", valueL)
    return valueL

def main():
    run = 1
    while run:
        r = input("Please enter the value of R: \n")
        if r.isdigit() and int(r) > 0:
            r = Decimal(r)
            run = 0
        else:
            print ("Please enter a valid integer value of R: \n")    

    calculateLength(Decimal(0.0001), r)
    
if __name__ == "__main__":
    main()
    

File: Concordia_COMP_6411_COMPARAT_STUDY_PROGRAM_LANG/diffCode/Javascript/POTUM.js
/* 
 *Filename:POTUM.JS
 *Author: Aravindan Balasubramanian
 *Team: TEAM - D
 *Course Code : COMP 6411
 *Date: 07/11/13
 *Description: The Beverage Coaster Project: Finds the length of overlap of two beverage coasters
  such that all the three sections have same Area A/2.
  
  To run this Implementation, please read the Readme.txt (available in the source folder)

 */
     
        // Desired and Chosen Accuracy of calculations
        var accuracy;
         
        //alpha's exponent : powerNum
        var Num;
        var n;
         
        /*calculate the Factorial of the power*/
        function calculateFactorial(n){
          
            if(n<1){
                return 1;
            }
            else {
                return n*calculateFactorial(n-1);
            }   
        
          }
          
        /*Calculating the Pi by Chudnovsky Approximation */
        function calculatePi(){
            
            var Pi=0;
            var k =0;
            while (k < 28){
                
                var a =(Math.pow(-1,k)); 
                
                var b = (calculateFactorial(6*k));
                
                var c = (calculateFactorial(3*k));
                
                var d = (calculateFactorial(k));
                
                var e = Math.pow(d,3);
                
                var f = (13591409+545140134*k);
                
                var g = Math.pow(640320,3*k);
                
                
                Pi+=((a*b)/(c*e))*(f/g); 
                
                k++;
                
              } 
             
             Pi *= Math.sqrt(10005)/(4270934400);
             Pi = Math.pow(Pi,-1);
             
            var myDiv2=document.getElementById("d2");
            myDiv2.innerHTML=Pi;
            
            
            return Pi;  
            
        }  
        
        /*PowerNum = no.of terms to which the series is calculated. Decides Stopping point.*/      
        function calculatePowerNum(accuracy){     
            
            //powerNum = 3 as starting term alpha^3.Step size = 2 as  alpha^3,alpha^5,alpha^7..
           for(var powerNum=3;powerNum<1000;powerNum+=2){  
                
                //calculating Pi
                var calPi=calculatePi();
                
                //accuracy condition : (pi^k/k!)<accuracy
                if((Math.pow(calPi,powerNum)/(calculateFactorial(powerNum)))<accuracy){
                  //stop when accuracy reached  
                  break;
                }
                                         
             }  
            
             return powerNum;  
             
          }    
            
        /*Solve for alpha : a - sina = pi/2 using Taylor's Series for Sine */   
        function calculateAlpha(accuracy){
            
           //get the power upto which the series should be calculated
           var power = calculatePowerNum(accuracy);
           
           //get the calculated pi
           var calPi = calculatePi();
           
           // 2*pi/3 < alpha < pi 
           var alpha;
           
           for(alpha = 2*calPi/3;alpha<calPi;alpha+= accuracy){
             // seriesValue is the LHS of the above equation    
             var seriesValue=0;
             for(var n=1;n<=(power-1)/2;n++){
             // calculating the seriesValue upto a^13   

             seriesValue = seriesValue + (Math.pow(-1,n+1)*Math.pow(alpha,2*n+1))/
                        calculateFactorial(2*n+1);
                } 
               
                if(((calPi/2)-(seriesValue))<accuracy){//stop on reaching accuracy
                
                    break; 
                }
            }

            var myDiv6=document.getElementById("d6");
            myDiv6.innerHTML=alpha;
            return alpha;
            
        }      
        
        /*Calculate L using alpha and given Radius*/
        function calculateLength(accuracy,r){ 
            
            var R = r;
                         
            //get alpha
            var a = calculateAlpha(accuracy);          
            
            //get Pi
            var calPi = calculatePi();           
            
            //given: a - sin a = pi/2, calculate Sin Alpha 
            var sinAlpha = a - calPi/2;        
                   
            //Pythagorean trignometric Identity
            var cosAlpha = 0;
            cosAlpha= -1*(Math.sqrt((1-sinAlpha*sinAlpha)));            
            
            //Cos Half Angle Formula
            var cosHalfAlpha = Math.sqrt((1+cosAlpha)/2);           
            
            //Calculate the Overlap: X1X2 using given Eqn
            var lengthofOverlap = 2*R*(1-cosHalfAlpha);         
           
            return lengthofOverlap;
            
        }
         
        /*gets the radius from user*/    
        function  getRadius(accuracy){
            
            try {
                
                  var R = document.getElementById("rad").value ; 
                  // validate user input in the text box
                  if(R=="") throw "Error :empty box ; Please enter a value";
                  if(isNaN(R)) throw "Error: not a number; Please enter a number";
                  if(R<0) throw "Error: Radius cannot be negative; Please enter a positive number;";
                  if(R==0) throw "Error:  Radius cannot be zero ; Enter Radius greater than zero";
                  else throw "Accepted : Entered value is within range";
               }
               
            catch(err){
                 
                  var msg = document.getElementById("errormsg");
                  msg.innerHTML =" " + err + ".";
               }
                  if(R>0){
                       var L= calculateLength(accuracy,R);
                  }
                  var myDiv0=document.getElementById("d0");
                  myDiv0.innerHTML=L; 
                        
        }
             